flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTriggeringTest.java

```diff
@@ -283,31 +300,494 @@ public void testStopPeriodicScheduler() throws Exception {
+			0,
+			Integer.MAX_VALUE,
+			CheckpointRetentionPolicy.NEVER_RETAIN_AFTER_TERMINATION,
+			true,
+			false,
+			0);
+		CheckpointCoordinator coord = new CheckpointCoordinator(
+			new JobID(),
+			chkConfig,
+			new ExecutionVertex[] { vertex1 },
+			new ExecutionVertex[] { vertex1 },
+			new ExecutionVertex[] { vertex1 },
+			new StandaloneCheckpointIDCounter(),
+			new StandaloneCompletedCheckpointStore(1),
+			new MemoryStateBackend(),
+			Executors.directExecutor(),
+			manuallyTriggeredScheduledExecutor,
+			SharedStateRegistry.DEFAULT_FACTORY,
+			failureManager);
+
+		coord.startCheckpointScheduler();
+		// Periodic
+		final CompletableFuture<CompletedCheckpoint> onCompletionPromise1 = coord.triggerCheckpoint(
+			System.currentTimeMillis(),
+			CheckpointProperties.forCheckpoint(CheckpointRetentionPolicy.NEVER_RETAIN_AFTER_TERMINATION),
+			null,
+			true,
+			false);
+		assertTrue(coord.isTriggering());
+		assertEquals(0, coord.getTriggerRequestQueue().size());
+
+		// another trigger before the prior one finished
+		manuallyTriggeredScheduledExecutor.triggerPeriodicScheduledTasks();
+		assertTrue(coord.isTriggering());
+		assertEquals(1, coord.getTriggerRequestQueue().size());
+
+		manuallyTriggeredScheduledExecutor.triggerAll();
+		assertFalse(onCompletionPromise1.isCompletedExceptionally());
+		assertFalse(coord.isTriggering());
+		assertEquals(0, coord.getTriggerRequestQueue().size());
+	}
+
+	@Test
+	public void testTriggerCheckpointRequestQueuedWithFailure() throws Exception {
+		// create some mock Execution vertices that receive the checkpoint trigger messages
+		final ExecutionAttemptID attemptID1 = new ExecutionAttemptID();
+		final AtomicInteger taskManagerCheckpointTriggeredTimes = new AtomicInteger(0);
+		final SimpleAckingTaskManagerGateway.CheckpointConsumer checkpointConsumer =
+			(executionAttemptID,
+			jobId, checkpointId,
+			timestamp,
+			checkpointOptions,
+			advanceToEndOfEventTime) -> taskManagerCheckpointTriggeredTimes.incrementAndGet();
+		ExecutionVertex vertex1 = mockExecutionVertex(attemptID1, checkpointConsumer);
+
+		// set up the coordinator and validate the initial state
+		CheckpointCoordinatorConfiguration chkConfig = new CheckpointCoordinatorConfiguration(
+			600000,
+			600000,
+			0,
+			Integer.MAX_VALUE,
+			CheckpointRetentionPolicy.NEVER_RETAIN_AFTER_TERMINATION,
+			true,
+			false,
+			0);
+		CheckpointCoordinator coord = new CheckpointCoordinator(
+			new JobID(),
+			chkConfig,
+			new ExecutionVertex[] { vertex1 },
+			new ExecutionVertex[] { vertex1 },
+			new ExecutionVertex[] { vertex1 },
+			new TestingCheckpointIDCounter(),
+			new StandaloneCompletedCheckpointStore(1),
+			new MemoryStateBackend(),
+			Executors.directExecutor(),
+			manuallyTriggeredScheduledExecutor,
+			SharedStateRegistry.DEFAULT_FACTORY,
+			failureManager);
+
+		coord.startCheckpointScheduler();
+		// Periodic
+		final CompletableFuture<CompletedCheckpoint> onCompletionPromise1 = coord.triggerCheckpoint(
```

Comment:
nit: `onCompletionPromise1` -> `periodicCompletionPromise`?

Reply:
I think `onCompletionPromise1` is better here. It's not important whether it's periodic or not. The comment should be clearer.